### PR TITLE
Sections code for vector search 

### PR DESCRIPTION
**Pull request by Thejas-bhat**
- Introduces a new section called `faissVectorIndexSection` which maintains vector search related operations over a specific field - indexing a bunch of vectors and performing the similarity search on them. 
- The section implements the `section` interface from section.go which basically ensures that the vector indexing related stuff that happens during the segment's lifecycle is agnostic of any other section that's present (for eg inverted index for textual data).
- The section itself uses the https://github.com/blevesearch/faiss (a fork from the facebookreasearch's faiss library) library to do the actual work of building the vector index and performing search over it.
- Currently, two types of index - flat or IVFSQ is built and maintained based on the volume of vectors present for a field within a segment.
- Currently during search time, the underlying faiss index is consulted to get the similar vectors, constructs a vector postings list which is iterable and each element within it is a pair of docNum in which the similar vector is present it and the corresponding vector's score.
- Caution: Still under active development.

### CODE REVIEW

**Comment by abhinavdangeti** [comment]
https://github.com/blevesearch/bleve/pull/1869

File: section_vector_index.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
+//go:build densevector
+// +build densevector
```
Comment (abhinavdangeti) [inline section_vector_index.go:2]: rename tag to `vectors`

File: new.go [hunk 2]
```diff
@@ -267,7 +267,6 @@ func (s *interim) processDocument(docNum uint64,
 	result index.Document,
 	fieldLens []int, fieldTFs []index.TokenFrequencies) {
 	visitField := func(field index.Field) {
-
```
Comment (abhinavdangeti) [inline new.go:270]: Let's avoid unrelated code changes as much as possible.

File: section_vector_index.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+type vectorIndexSection struct {
+}
+
+func (v *vectorIndexSection) Process(opaque map[int]resetable, docNum uint64, field index.Field, fieldID uint16) {
+	if fieldID == math.MaxUint16 {
+		return
+	}
+
+	if vf, ok := field.(index.DenseVectorField); ok {
```
Comment (abhinavdangeti) [inline section_vector_index.go:27]: Recommend refactoring DenseVectorField -> VectorField (as commented in bleve_index_api).

File: section_vector_index.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	if cap(buf) < size {
+		buf = make([]byte, size)
+		v.tmp0 = buf
+	}
+	return buf[0:size]
+}
+
+func (vo *vectorIndexOpaque) writeVectorIndexes(w *CountHashWriter) (offset uint64, err error) {
+	// for every fieldID, contents to store over here are:
+	//    1. the serialized representation of the dense vector index.
```
Comment (abhinavdangeti) [inline section_vector_index.go:72]: Drop references to "dense" where applicable.

File: section_vector_index.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+		vo.allocateSpace()
+	}
+	if fieldID == math.MaxUint16 {
+		// doc processing checkpoint. currently nothing to do
+		return
+	}
+
+	//process field
+
+	vec := field.DenseVector()
```
Comment (abhinavdangeti) [inline section_vector_index.go:184]: API name refactor: DenseVector -> Vector.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: go.mod [hunk 6]
```diff
@@ -19,3 +20,9 @@ require (
 	github.com/spf13/pflag v1.0.5 // indirect
 	golang.org/x/sys v0.0.0-20220520151302-bc2c85ada10a // indirect
 )
+
+replace github.com/blevesearch/bleve_index_api => ../bleve_index_api
```
Comment (metonymic-smokey) [inline go.mod:24]: Nit: pls remove these replace directives.

Comment (metonymic-smokey) [reply go.mod:24]: I'm guessing you added them for local testing.

File: section_vector_index.go [hunk 7]
```diff
@@ -0,0 +1,458 @@
+//go:build vectors
```
Comment (abhinavdangeti) [inline section_vector_index.go:1]: Add that copyright snippet above here.

**Review by abhinavdangeti — COMMENTED** [review]
Looking forward to a lot of commentary here :)

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**